src/Processors/Streaming/ISource.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <Checkpoint/CheckpointRequest.h>
+#include <Processors/ISource.h>
+#include <Common/serde.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int NOT_IMPLEMENTED;
+}
+
+namespace Streaming
+{
+class ISource : public DB::ISource
+{
+public:
+    ISource(Block header, bool enable_auto_progress, ProcessorID pid_) : DB::ISource(std::move(header), enable_auto_progress, pid_)
+    {
+        is_streaming = true;
+    }
+
+    virtual String description() const { return ""; }
+
+    /// \brief Get the last progressed sequence number of the source, it shouldn't be called during pipeline execution (thread-unsafe)
+    virtual Int64 lastSN() const { throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented for lastSN() of {}", getName()); }
+
+    /// \brief Reset the start sequence number of the source, it must be called before the pipeline execution (thread-unsafe)
+    void resetStartSN(Int64 sn);
+
+    Int64 lastCheckpointedSN() const noexcept { return last_checkpointed_sn; }
+
+    void checkpoint(CheckpointContextPtr ckpt_ctx_) override final;
```

Comment:
can we make `checkpoint` non virtual and remove the `checkpoint, recover` interface from DB::ISource?